### ISSUE TITLE
Rails New Framework Default Documentation is Inadequate

### ISSUE THREAD

**Issue by BenSandeen** [rails/rails#50238]
### Steps to reproduce
Go to https://guides.rubyonrails.org/v7.0/configuring.html and try to figure out what each setting does

### Expected behavior
Documentation documents things and explains how they work

### Actual behavior
Documentation restates each new framework default in plain English, typically with little or no additional information

### System configuration
**Rails version**: 7.0

**Ruby version**: 3.0

Sorry for being a bit flippant, but it's _incredibly frustrating_ when trying to turn on these new framework defaults (which are, by definition, very likely to be breaking changes) and the official documentation doesn't explain the before and after states or give examples of the impact the new config might have.  To me, it feels like going to swimming lessons and all the instructor says is, "Keep your head above the water."

Sorry if this isn't the right place to report this problem.

**Comment by fatkodima** [comment]
Can you provide an example of the configuration from the docs that you had trouble understanding? And describe a bit what that trouble was.

**Comment by BenSandeen** [comment]
At least half of the new framework defaults left me frustratingly under-informed.  If you'd really like, I can spend a couple of hours and go through each one and list all the unanswered questions each leaves me with, but it's really more of a systemic question rather than a few bad examples here and there

**Comment by skipkayhil** [comment]
> At least half of the new framework defaults left me frustratingly under-informed.

Can you understand how this isn't actionable feedback from our perspective? If we don't know what isn't documented well enough, it's hard to know what to actually change

> If you'd really like, I can spend a couple of hours and go through each one

Please do, if the documentation is this bad then we should take the time to fix it

**Comment by BenSandeen** [comment]
> > At least half of the new framework defaults left me frustratingly under-informed.
> 
> Can you understand how this isn't actionable feedback from our perspective? If we don't know what isn't documented well enough, it's hard to know what to actually change

I can totally understand!  😃 that's why I waited so long to submit this issue, because I wasn't sure if this was the best place to submit the feedback

> > If you'd really like, I can spend a couple of hours and go through each one
> 
> Please do, if the documentation is this bad then we should take the time to fix it

I'll start working on it, but it'll probably take a few days at the bare minimum for me to do this in between all the other stuff at work

**Comment by skipkayhil** [comment]
> I'll start working on it, but it'll probably take a few days at the bare minimum for me to do this in between all the other stuff at work

No rush at all! Take your time and feel free to provide smaller updates instead of feeling like you have to do them all at once 🙂

**Comment by BenSandeen** [comment]
All right, here's my really basic write-up of each of the Rails 7.0 new framework defaults (because the 6.1 to 7.0 upgrade process is the one I'm currently struggling through)

To create this document, I basically copied and pasted the new framework default names from https://guides.rubyonrails.org/v7.0/configuring.html#default-values-for-target-version-7-0.  I then copied and pasted the short description of the corresponding new framework default, followed by any comments or questions the description leaves me with.

Keep in mind that I'm a somewhat junior backend software engineer.  Frontend engineers would probably have fewer questions on some of the more frontend-y things.

## `config.action_controller.raise_on_open_redirects`

> Raises an `ArgumentError` when an unpermitted open redirect occurs.

- What is an open redirect?
- What determines whether or not an open redirect is permitted?

## `config.action_controller.wrap_parameters_by_default`

> Configures the `ParamsWrapper` to wrap json request by default.

- How does it wrap JSON requests?
  + Examples of JSON requests with this NFD set to true vs false (or before and after Rails 7.0) would be extremely helpful
- Why is wrapping JSON requests desirable?
- The comment in the `new_framework_defaults_7_0.rb` file is extremely confusing
```
# Enable parameter wrapping for JSON.
# Previously this was set in an initializer. It's fine to keep using that initializer if you've customized it.
# To disable parameter wrapping entirely, set this config to `false`.
```
- The line that says "Previously this was set in an initializer. It's fine to keep using that initializer if you've customized it" is confusing; what should I set the new framework default to if I do want to keep using an existing initializer?  It sounds like it's telling us to keep doing what we were doing before Rails 7.0, implying that we should use the Rails 6.1 default of `false`
- However, the next line, "To disable parameter wrapping entirely, set this config to `false`" pretty clearly indicates that we don't want to set it to `false` if we want to keep pre-existing parameter wrapping

## `config.action_dispatch.cookies_serializer`

> Specifies which serializer to use for cookies. For more information, see Action Controller Cookies.

- This is probably fine, since it at least gives a link to an explanation of how Rails handles cookies

## `config.action_dispatch.default_headers`

- This one is great.  It shows exactly the difference between the old and new versions.  Searching online for explanations of the differences is quite easy

## `config.action_dispatch.return_only_request_media_type_on_content_type`

> Change the return value of ActionDispatch::Request#content_type to the Content-Type header without modification.

- What does the "Content-Type" header look like?
- What does the "Content-Type" header look like without modifications?
- Thankfully, this section on a completely different page sufficiently documents the change: https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#actiondispatch-request-content-type-now-returns-content-type-header-as-it-is
  + It would be ideal to just have the change in the NFD documentation page (at the very least, there ought to be a link from the NFD page to the URL I included above)
  
## `config.action_mailer.smtp_timeout`

> Allows to configure both the `:open_timeout` and `:read_timeout` values for `:smtp` delivery method.

- Some more information would be nice, but the single sentence at least contains sufficient information to make searching online feasible

## `config.action_view.apply_stylesheet_media_default`

> Determines whether `stylesheet_link_tag` will render `screen` as the default value for the attribute `media` when it's not provided.

- Some more information would be nice, but the single sentence at least contains sufficient information to make searching online feasible
- This article has a great explanation: https://dev.to/thomasvanholder/rails-7-framework-defaults-part-1-2a59#2-applystylesheetmediadefault

## `config.action_view.button_to_generates_button_tag`

> Determines whether `button_to` will render `<button>`` element, regardless of whether or not the content is passed as the first argument or as a block.

- What does this look like in practice?
- Some sample code showing the output of `button_to` with this NFD turned off and on would be extremely helpful
  + This article has a great and simple explanation: https://dev.to/thomasvanholder/rails-7-framework-defaults-part-1-2a59#1-buttontogeneratesbuttontag

## `config.active_record.automatic_scope_inversing`

> Enables automatically inferring the `inverse_of` for associations with a scope.

- Some more specific information would be greatly appreciated
  + Particularly, an explanation about why this is considered a potential breaking change would be helpful, because it doesn't seem like it should cause much of anything to break; it just sounds like a nice new feature to make working with models easier
- Here are some helpful articles that I found
  + https://blog.kiprosh.com/rails-7-auto-detects-inverse-of-for-scoped-associations/
  + https://blog.saeloun.com/2022/02/01/rails-7-inverse-of-automatic-inference/
  + https://rubyonrails.org/2021/10/10/this-week-in-rails-automatic-inverse_of-performance-improvements-and-more

## `config.active_record.partial_inserts`

> Is a boolean value and controls whether or not partial writes are used when creating new records (i.e. whether inserts only set attributes that are different from the default).

- The "Is a boolean value and" bit is superfluous (it's not a problem to have it, but it's not entirely necessary considering the true/false being listed as the before and after values)
- The explanation isn't great, but it includes enough information to make searching online for an actual explanation possible
- Here's a great explanation I found: https://blog.kiprosh.com/rails-7-introduces-partial-inserts-config-for-activerecord/
  + It would have been extremely helpful to include example SQL queries with this turned off and on right in the NFD documentation
- Also, a brief explanation of why this was turned off in Rails 7.0 would have been nice, since it is a clear loss of optimization
  + This Rails PR has a decent and quick explanation: https://github.com/rails/rails/pull/42769

## `config.active_record.verify_foreign_keys_for_fixtures`

> Ensures all foreign key constraints are valid after fixtures are loaded in tests. Supported by PostgreSQL and SQLite only.

- This is mostly good
- An example would be helpful, like here: https://dev.to/thomasvanholder/rails-7-framework-defaults-part-1-2a59#3-verifyforeignkeysforfixtures

## `config.active_storage.multiple_file_field_include_hidden`

> In Rails 7.1 and beyond, Active Storage `has_many_attached` relationships will default to replacing the current collection instead of appending to it. Thus to support submitting an empty collection, when `multiple_file_field_include_hidden` is `true`, the `file_field` helper will render an auxiliary hidden field, similar to the auxiliary field rendered by the `check_box` helper.

- Why is the "auxiliary hidden field" necessary or helpful?
  + What does it actually look like in code?
  + Why does it "support submitting an empty collection"?
- Why does Rails 7.1's new default for `has_many_attached` necessitate supporting the submission of empty collections?

## `config.active_storage.variant_processor`

> Accepts a symbol `:mini_magick` or `:vips`, specifying whether variant transformations and blob analysis will be performed with MiniMagick or ruby-vips.

- This is more or less sufficient
- It would be nice to have a couple of links to more information detailing how to use MiniMagick and ruby-vips

## `config.active_storage.video_preview_arguments`

> Can be used to alter the way ffmpeg generates video preview images.

- The provided before and after default values are sufficient to understand more or less what this means
- (We don't use Rails's ActiveStorage, so maybe I wouldn't have this question if my company used it): How do we tell ffmpeg to generate video preview images?  Does that change at all?

## `config.active_support.cache_format_version`

> Specifies which version of the cache serializer to use. Possible values are 6.1 and 7.0.

- What effects does this have?
- Is this reversible?
  + I know the comment in the `new_framework_defaults.rb` file answer both of these questions a bit, so it's not a critical omission, but it feels odd to not mention the effects in the documentation as well

## `config.active_support.disable_to_s_conversion`

> Disables the override of the `#to_s` methods in some Ruby core classes. This config is for applications that want to take advantage early of a Ruby 3.1 optimization. This configuration needs to be set in `config/application.rb` inside the application class, otherwise it will not take effect.

- Which Ruby core classes does it affect?
- These release notes seem relevant, but the messaging is different: https://github.com/rails/rails/blob/de7c495209811f8f918fa9f915e64df61a6080c1/guides/source/7_0_release_notes.md#deprecations-8
  + Is the list of classes here the entirety of the classes that the documentation refers to?

## `config.active_support.executor_around_test_case`

> Configure the test suite to call `Rails.application.executor.wrap` around test cases. This makes test cases behave closer to an actual request or job. Several features that are normally disabled in test, such as Active Record query cache and asynchronous queries will then be enabled.

- A more exhaustive list of the specific things that change would be fantastic

## `config.active_support.hash_digest_class`

> Allows configuring the digest class to use to generate non-sensitive digests, such as the ETag header.

- What's a digest?
- What is the digest class?
- What makes a digest sensitive?
- What does the "such as the ETag header" part mean?  (The sentence is doesn't quite make sense)
- What effects will changing this setting have?

## `config.active_support.remove_deprecated_time_with_zone_name`

> Specifies whether to remove the deprecated override of the `ActiveSupport::TimeWithZone.name` method, to avoid triggering its deprecation warning.

- Why is this deprecation kept behind a new framework default?  (What sort of breaking change could it lead to?)
- What kind of effects would leaving it in have?
- What if future versions of Rails remove the deprecated method but someone still has this new framework default set to `false`?  (Will things just break?)

## `config.active_support.use_rfc4122_namespaced_uuids`

> Specifies whether generated namespaced UUIDs follow the RFC 4122 standard for namespace IDs provided as a String to Digest::UUID.uuid_v3 or Digest::UUID.uuid_v5 method calls.
> If set to true:
> 
>   * Only UUIDs are allowed as namespace IDs. If a namespace ID value provided is not allowed, an ArgumentError will be raised.
>   * No deprecation warning will be generated, no matter if the namespace ID used is one of the constants defined on Digest::UUID or a String.
>   * Namespace IDs are case-insensitive.
>   * All generated namespaced UUIDs should be compliant to the standard.
>
> If set to false:
>
>   * Any String value can be used as namespace ID (although not recommended). No ArgumentError will be raised in this case in order to preserve backwards-compatibility.
>   * A deprecation warning will be generated if the namespace ID provided is not one of the constants defined on Digest::UUID.
>   * Namespace IDs are case-sensitive.
>   * Only namespaced UUIDs generated using one of the namespace ID constants defined on Digest::UUID are compliant to the standard.

- I like that it's explicit about what impacts `true` and `false` means!
- What are namespace IDs?
- I found these two PRs to be helpful in supplementing my understanding:
  + https://github.com/rails/rails/issues/37681
  + https://github.com/rails/rails/pull/37682

**Comment by erichmachado** [comment]
> * What are namespace IDs?

I can provide more context there, if it helps. However, it seems like the entire documented section [was removed](https://github.com/rails/rails/commit/7b4affc78bf2fdd349a86c60b940b7c172e111df#diff-210d0f245e618b7f8c6fcb53f518bd7fd343ce7e5d8c609ecfdb149418beca19) from the latest [`guides/source/configuring.md`](https://github.com/rails/rails/blob/v7.1.2/guides/source/configuring.md) when this option was deprecated in the latest release.

I'm not sure if this needs to be addressed or how to proceed here. /cc @skipkayhil

**Comment by bparanj** [comment]
In Ruby on Rails, configuration settings are crucial for defining the behavior of your application in different environments (development, test, and production). These settings are typically organized in specific files within the `config` directory. Here's a breakdown:

1. **config/environments/**: This directory contains environment-specific settings. There are typically three files: `development.rb`, `test.rb`, and `production.rb`. Each file contains settings that are only relevant to that environment. For example, `development.rb` might have configurations for debugging and reloading code, `test.rb` for test suite settings, and `production.rb` for performance optimizations and security settings.

2. **config/database.yml**: This file configures the database connection. You can define settings for different environments, specifying which database to use (e.g., SQLite, PostgreSQL), along with credentials and other database-specific settings.

3. **config/routes.rb**: This file defines the URL routing for your application. It maps incoming requests to controllers and actions.

4. **config/secrets.yml**: This file is used to manage secret keys. It's important for security, storing credentials like API keys and database passwords. Rails 5.2 introduced `credentials.yml.enc` as a new way to handle secrets, encrypting them for added security.

5. **config/application.rb**: This file is used to configure settings that apply to all environments, such as timezone, locale, and loading paths for Rails components.

6. **config/initializers/**: This directory contains small configuration files that set up various parts of your Rails application. Files here are executed when the application starts.

7. **config/locales/**: This directory is for internationalization, containing files that define text strings in different languages.

8. **config/boot.rb**: This file is used for booting your application. It's typically not altered much but is crucial for the startup process.

9. **config/cable.yml**: If you're using Action Cable (WebSocket framework in Rails), this file is used to configure its settings.

10. **config/credentials.yml.enc**: Introduced in Rails 5.2, this file securely stores encrypted credentials, accessed via `Rails.application.credentials`.

11. **config/master.key** or **config/credentials/production.key**: These are key files used to decrypt `credentials.yml.enc`. They should be kept secure and not checked into version control.

Each of these configurations serves a distinct purpose, helping to modularize and organize the settings for different aspects of your Rails application. Understanding and managing these configurations is key to maintaining a robust, secure, and efficient Rails application.

## Config Application

The `config/application.rb` file in a Ruby on Rails application plays a crucial role in configuring settings that are global to your application, affecting all environments (development, test, and production). Here's a detailed explanation of its purposes and common uses:

1. **Application Settings**: It sets up defaults for your Rails application. This includes configuration for:
   - Time zone (`config.time_zone`).
   - The default locale (`config.i18n.default_locale`).
   - The autoload paths, which determines where Rails looks for files to load.
   - The load paths for helpers, enabling or disabling them across all controllers.

2. **Middleware Stack Configuration**: Rails applications use a middleware stack that you can modify in this file. You can add, remove, or configure middleware components, which are executed for every incoming HTTP request.

3. **Generators Configuration**: You can customize the behavior of Rails generators. This includes specifying which test framework, JavaScript framework, or stylesheet engine to use. For example, you can tell Rails to generate RSpec tests instead of the default Minitest by configuring the test framework.

4. **Active Record Configuration**: While specific database settings are in `database.yml`, `application.rb` can be used to set global Active Record behaviors. For example, you can configure the schema format (Ruby or SQL format) for database migrations.

5. **Action Dispatch Configuration**: This includes settings for the routing aspect of Rails. You can, for example, configure custom path parameters or constraints.

6. **Asset Pipeline Configuration**: If you are using the asset pipeline, `application.rb` is where you configure its behavior. This includes settings like the assets version and adding paths to the asset search path.

7. **Action Mailer Configuration**: Global settings for Action Mailer, such as default URL options and mailer layouts, are specified here.

8. **Logging Configuration**: You can define the logging level (e.g., debug, info, warn, error) and customize the formatting of logged messages.

9. **Framework Defaults**: It's also common to set or override various defaults for Rails components, like Action Controller, Action View, etc.

10. **Initialization Events**: Rails offers hooks like `before_configuration`, `before_initialize`, and `after_initialize` that you can use to execute code at specific points in the application boot process.

The `application.rb` file essentially acts as the central place for high-level configuration of your Rails application. Properly setting up and understanding this file is crucial for defining how your application behaves across various environments and components.

**Comment by bparanj** [comment]
## How to find all the config available in Rails.

To find all available configuration options from the Rails console, you can use various inspection methods provided by Rails. Here's a step-by-step guide to do this:

1. **Start the Rails Console**: 
   - Open your terminal.
   - Navigate to your Rails application directory.
   - Run the command `rails console` or `rails c` to start the console.

2. **Inspect Application Configurations**:
   - Once in the console, you can inspect different configuration objects. For example, to see all the configurations related to your application, you can use `Rails.application.config`.

3. **Specific Configurations**:
   - For inspecting specific parts of the configuration, like the database, environments, or middleware, you can use targeted commands. For instance:
     - `Rails.application.config.database_configuration` will display the database configuration.
     - `Rails.application.config_for(:database)` is another way to access database configuration.
     - `Rails.application.config.middleware` will show you the middleware stack.
     - `Rails.application.config.active_record` will display Active Record configurations.
     - `Rails.application.config.action_mailer` for Action Mailer configurations.

4. **Use `inspect` or `to_h` Methods**:
   - To get a more readable output, you can append `.inspect` or `.to_h` (to convert to a hash) to these commands. For example, `Rails.application.config.to_h`.

5. **Exploring Environment-Specific Configurations**:
   - If you want to explore configurations for a specific environment, you can access them via `Rails.application.config_for(:environment_name)`. Replace `:environment_name` with `:production`, `:development`, or `:test` as needed.

6. **Active Support Configurable**:
   - Rails uses the ActiveSupport::Configurable module in many components. You can explore configurations of individual components like Action Controller, Active Record, etc., using similar commands.

7. **Rails Guides and API Documentation**:
   - For a comprehensive list of configuration options, the Rails Guides and API documentation are excellent resources. They provide detailed information about various configuration options available in Rails.

Remember, the Rails console is a powerful tool for inspecting and interacting with your application. It can be used not only to check configurations but also to test methods, models, and more in the Rails environment.

**Comment by bparanj** [comment]
The scope of this request is too big. Can we close this issue?

**Comment by rafaelfranca** [comment]
> What is an open redirect?

Google can answer that.

> What determines whether or not an open redirect is permitted?

This definitely can be explained in the guide. PRs welcome.

> How does it wrap JSON requests?
That is on the documentation of `ParamsWrapper` https://api.rubyonrails.org/classes/ActionController/ParamsWrapper.html

> Why is wrapping JSON requests desirable?
It is up to you. If you want, do it, if you don't care, do nothing. That is why we are providing a default.

> The line that says "Previously this was set in an initializer. It's fine to keep using that initializer if you've customized it" is confusing; what should I set the new framework default to if I do want to keep using an existing initializer? 

To nothing. If you want to keep the existing initializer you can just delete this default, you are already setting the same config in the initializer.

> What does the "Content-Type" header look like?
You can find this on google, but well, like any "Content-Type" header in any web page looks like

I'm not going to answer all the points, and I totally understand the frustration, I do, but when I read those questions I can't help to think that there is some knowledge missing on how web applications work, and the Rails guides at the moment has no intention to teach people how the web works. If that is the case, you should not change any of the frameworks default.

I don't disagree we can document those configurations better, and for that, pull request welcome, but I have to agree with @bparanj, this issue as it is is to broad. I'm happy to receive PRs to improve the documentation. I'm happy to discuss any specific doubts you have about those configs and Rails upgrades in general in the category in the forum https://discuss.rubyonrails.org/c/rails-upgrades/11, but I don't think leaving this issue open will lead anywhere.

**Comment by jgarber623** [comment]
@BenSandeen Thank you for the links in [your comment](https://github.com/rails/rails/issues/50238#issuecomment-1841562696). They're quite helpful!

**Comment by wwahammy** [comment]
I appreciate why this was closed but thank you @BenSandeen. Your review is one of the best places to actually understand what these defaults mean for Rails 7. Can you do this for every version going forward? :)

**Comment by BenSandeen** [comment]
> I appreciate why this was closed but thank you @BenSandeen. Your review is one of the best places to actually understand what these defaults mean for Rails 7. Can you do this for every version going forward? 🙂

@wwahammy Ha, I'm glad this helped you!  I certainly can't commit to to doing this in the future, but I'll keep it in mind.  And for what it's worth, I'm working through the upgrade from 7.0 to 7.1 on a couple of smaller repos, and the documentation is a fair bit less bad (I still can't say it's great, but at least it doesn't feel like it's trying to hide useful information in order to make your life harder) 😛
